docs/proposals/block-storage-time-series-deletion.md

```diff
@@ -0,0 +1,247 @@
+
+Pros:
+- Simpler implementation as everything is contained within the DeletedSeriesCleaner.
+
+Cons:
+- Might have to wait for a longer period of time for the compaction to be finished. 
+    - This would mean the earliest time to be able to run the deletion would be once the last time from the block_ranges in the [compactor_config](https://cortexmetrics.io/docs/blocks-storage/compactor/#compactor-configuration) has passed. By default this value is 24 hours, so only once 24 hours have passed and the new compacted blocks have been created, then the rewrite can be safely run. 
+
+
+
+
+Option 2: For blocks that still need to be compacted further after the deletion request cancel period is over, the deletion logic can be applied before the blocks are compacted. This will generate a new block which can then be used instead for compaction with other blocks. 
+
+Pros:
+- The deletion can be applied earlier than the previous options. 
+    - Only applies if the deletion request cancel period is less than the last time interval for compaction is. 
+Cons:
+- Added coupling between the compaction and the DeletedSeriesCleaner. 
+- Might block compaction for a short time while doing the deletion. 
+
+
+
+Once all the applicable blocks have been rewritten without the deleted data, the deletion request state moves to `Syncing`. Once the syncing time period is over, the state will advance to `Processed` state and the tombstone will no longer be used.  
+
+
+
+#### Handling failed/unfinished delete jobs:
+
+Deletions will be completed and the tombstones will be deleted only when the DeletedSeriesCleaner iterates over all blocks that match the time interval and confirms that they have been re-written without the deleted data.  Otherwise, it will keep creating the markers indicating which blocks are remaining for deletion. In case of any failure that causes the deletion to stop, any unfinished deletions will be resumed once the service is restarted. The series deletion markers will remain in the bucket until the new blocks are created without the deleted data. Meaning that the compactor will continue to process the blocks for deletion that are remaining according to the deletion markers. 
+
+
+#### Tenant Deletion API
+
+If a request is made to delete a tenant, then all the tombstones will be deleted for that user. 
+
+
+
+## Current Open Questions:
+
+- If the start and end time is very far apart, it might result in a lot of the data being re-written. Since we create a new block without the deleted data and mark the old one for deletion, there may be a period of time with lots of extra blocks and space used for large deletion queries. 
+- There will be a delay between the deletion request and the deleted data being filtered during queries. 
+    - In Prometheus, there is no delay. 
+    - One way to filter out Immediately is to load the tombstones during query time but this will cause a negative performance impact. 
```

Comment:
Assuming query-frontends know "cache gen num", they could tell it to queriers, which can then reload deletion markers based on that. [But this breaks when using bucket-index, which we want, so it's not a good idea].